File: linguagem_c/cliente.c
#include "cliente.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>

typedef struct cliente {
    char nome[100];
    char cidade[100];
    char codigo[100];
} Cliente;

void cabecalho(char* linha, char* titulo) {
    printf("%s\t%s\t%s\n", linha, titulo, linha);
}

void pressiona_enter(){
    printf("\nPressione ENTER para continuar.");
    while (getchar() != '\n');
    getchar();
}

int numero_inteiroc(char *str) {
    for (int i = 0; str[i] != '\0'; i++) {
        if (!isdigit(str[i])) {
            return 0;
        }
    }
    return 1;
}

void formata_string(char *str) {
    int i;
    for (i = 0; str[i] != '\0'; i++) {
        if (i==0){
            str[0] = toupper(str[0]);
        }
        else{
            if (str[i - 1] == ' ') {
                str[i] = toupper(str[i]);
            } 
            else {
                str[i] = tolower(str[i]);
            }
        }
    }
}

int verifica_codigo(char *codigo_digitado) {
    char linha[300];
    char codigo_arquivo[100];
    char cidade_arquivo[100];
    char nome_arquivo[100];

    FILE *arquivo = fopen("clientes.txt", "r"); 
    if (arquivo == NULL) {
        return 0; 
    }

    while (fgets(linha, sizeof(linha), arquivo) != NULL) { 
        if (sscanf(linha, "%99[^\t]\t%99[^\t]\t%99[^\t ]", nome_arquivo, cidade_arquivo, codigo_arquivo) == 3) {
            if (strcmp(codigo_arquivo, codigo_digitado) == 0) {
                printf("\nCodigo ja existe.");
                fclose(arquivo);
                return 1; 
            }
        }
    }

    fclose(arquivo);
    return 0; 
}

void adiciona_cliente() {
    Cliente cliente;
    FILE *arquivo = fopen("clientes.txt", "a"); 
    if(arquivo == NULL) { 
        printf("Erro ao abrir o aquivo.\n");
        return;
    }

    char nome_digitado[100];
    char cidade_digitada[100];
    char codigo_digitado[100]; 


    cabecalho("============", "Adicionar Cliente");
    do{
    printf("\nDigite o nome do cliente (apenas letras): ");
    scanf("%99[^\n]", nome_digitado);
    getchar();
    }while (!contem_apenas_letras(nome_digitado));
    
    formata_string(nome_digitado);
    strcpy(cliente.nome, nome_digitado);

    do{
        printf("\nDigite a cidade do cliente (apenas letras): ");
        scanf("%99[^\n]", cidade_digitada);
        getchar();
    }while (!contem_apenas_letras(cidade_digitada));
        
    formata_string(cidade_digitada);
    strcpy(cliente.cidade, cidade_digitada);

    do{
    printf("\nDigite o codigo do cliente (apenas numeros inteiros): ");
    scanf("%99[^\n]", codigo_digitado);
    getchar();
    }while (!numero_inteiroc(codigo_digitado) || verifica_codigo(codigo_digitado));

    strcpy(cliente.codigo, codigo_digitado);

    fprintf(arquivo, "\n%s \t %s \t %s \n", cliente.nome, cliente.cidade, cliente.codigo);

    fclose(arquivo);
}

void procura_codigo() {
    char codigo_digitado[100];
    char linha[300];

    FILE *arquivo = fopen("clientes.txt", "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo.\n");
        return;
    }

    do{
    printf("\nDigite o codigo do cliente (apenas numeros inteiros): ");
    scanf("%99[^\n]", codigo_digitado);
    getchar();
    }while (!numero_inteiroc(codigo_digitado));
  
    int encontrou = 0;
    while (fgets(linha, sizeof(linha), arquivo)) {
        char nome[100], cidade[100], codigo_arquivo[100];
        if (sscanf(linha, "%99[^\t]\t%99[^\t]\t%99[^\t ]", nome, cidade, codigo_arquivo) == 3) {
            if (strcmp(codigo_arquivo, codigo_digitado) == 0) {
                printf("Cliente encontrado\n");
                printf("Nome: %s\nCidade: %s\nCodigo: %s\n", nome, cidade, codigo_arquivo);
                encontrou = 1;
                break;
            }
        }
    }

    if (!encontrou) {
        printf("Cliente nao encontrado!\n");
    }

    pressiona_enter();

    fclose(arquivo);
}

int contem_apenas_letras(char *str) {
    for (int index = 0; str[index] != '\0'; index++) {
        if (!isalpha(str[index]) && str[index] != ' ') {
            printf("A string deve conter apenas letras.\n");
            return 0; 
        }
    }
    return 1; 
}

void procura_nome() {
    char nome_digitado[100];
    char linha[300];

    FILE *arquivo = fopen("clientes.txt", "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo.\n");
        return;
    }

    do{
        printf("Digite o nome do cliente que deseja buscar: ");
        scanf(" %99[^\n]", nome_digitado);
    } while (!contem_apenas_letras(nome_digitado));
    int encontrou = 0;

    formata_string(nome_digitado);

    while (fgets(linha, sizeof(linha), arquivo)) {
        char nome_arquivo[100], cidade[100], codigo[100];
        if (sscanf(linha, "%99[^\t]\t%99[^\t]\t%99[^\t ]", nome_arquivo, cidade, codigo) >= 2) {
            if (strcmp(nome_arquivo, nome_digitado) == 0 || strstr(nome_arquivo, nome_digitado) != NULL) {
                printf("Nome: %s\nCidade: %s\nCodigo: %s\n", nome_arquivo, cidade, codigo);
                encontrou = 1;
            }
        }
    }

    if (!encontrou) {
        printf("Cliente nao encontrado!\n");
    }

    pressiona_enter();
    fclose(arquivo);
}

void lista_clientes(){
    FILE *arquivo = fopen("clientes.txt", "r"); 
    char linha[300];

    if(arquivo == NULL) {
        printf("Erro ao abrir o aquivo.\n");
        return;
    }
    
    cabecalho("====", "Lista de clientes");
    printf("%s \t\t %s \t %s\n", "Nome", "Cidade", "Codigo");

    while(fgets(linha, sizeof(linha), arquivo)) {
        printf("%s", linha); 
    }
    
    pressiona_enter();

    fclose(arquivo);

} 

void menu() {
    char opcao[100];

    char sistema_operacional[10];
    strcpy(sistema_operacional, "cls");

    while (1) {
        cabecalho("============", "Menu principal");

        printf("1|\t Adicionar cliente\n");
        printf("2|\t Listar clientes\n");
        printf("3|\t Procurar cliente por codigo\n");
        printf("4|\t Procurar cliente por nome\n");
        printf("5|\t Sair\n");
        printf("---------------------------------------------\n");

        do{
        printf("Escolha uma opcao dentre as do menu: ");
        scanf(" %[^\n]", opcao);
        getchar();
        }while(!numero_inteiroc(opcao));

        if (strcmp(opcao, "1") == 0) {
            system(sistema_operacional); 
            adiciona_cliente();
            sleep(1); 
            system(sistema_operacional); 
        } else if (strcmp(opcao, "2") == 0) {
            system(sistema_operacional);
            lista_clientes();
            system(sistema_operacional);
        } else if (strcmp(opcao, "3") == 0) {
            system(sistema_operacional);
            procura_codigo();
            system(sistema_operacional);
        } else if (strcmp(opcao, "4") == 0) {
            system(sistema_operacional);
            procura_nome();
            system(sistema_operacional);
        } else if (strcmp(opcao, "5") == 0) {
            printf("Saindo...\n"); 
            sleep(1);
            system(sistema_operacional);
            break;
        } else {
            printf("Opcao invalida!\n");
            sleep(1);
            system(sistema_operacional);
        }
    }
}

File: linguagem_python/cliente.py
import os
from time import sleep
SISTEMA_OPERACIONAL = "cls" #definido para windows

def cabeçalho(divisoria, titulo):
    """Imprime um cabeçalho com um título centralizado
    
    Args:
        divisoria (str): Caractere que será repetido 50 vezes
        titulo (str): Título que será centralizado
    Returns:
        None    
    """
    print(divisoria * 50)
    print(f"{titulo:^50}")
    print(divisoria * 50)

def valida_codigo(verificar_existencia):
    """Valida o código do cliente, vendo se é um número, além de verificar se já existe um cliente com esse código
    de acordo com o parâmetro passado.

    Args:
        verificar_existencia (bool): Parâmetro que indica se a função deve verificar se o código já existe.

    Returns:
        int: Código do cliente válido
    """
    
    while True:
        try:
            codigo = int(input("Digite o código de cliente: "))
            break
        except ValueError:
            print('Insira um código válido')

    if verificar_existencia:
        while True:
            if verifica_codigo(codigo):
                print("Código já existente!")
        
                while True:
                    try:
                        codigo = int(input("Digite o código de cliente: "))
                        break
                    except ValueError:
                        print('Insira um código válido')
            else:
                break
            
    return codigo

def valida_string(mensagem):
    """Valida uma string, verificando se é vazia ou se contém apenas espaços em branco. 
    Também verifica se contém números ou caracteres especiais.
    
    Args:
        mensagem (str): Mensagem que será exibida ao usuário
    Returns:
        string: string válida e formatada com a primeira letra de cada palavra em maiúsculo.
    """
    while True:
        string = input(mensagem)
        if string.strip() == "":
            print("Digite algo válido!")
        elif not ((string).replace(" ", "")).isalpha():
            print("Não é permitido números ou caracteres especiais!")
        else:
            break
    
    return string.title()

def adiciona_cliente():
    """Adiciona um cliente ao arquivo clientes.txt
    
    Args:
        None
    Returns:
        None
    """

    arquivo = open('clientes.txt', 'a')
    
    cliente = {}
    
    cabeçalho("-", "Adicionar cliente")
    
    nome = valida_string("Digite o nome do cliente: ")
    cidade = valida_string("Digite a cidade do cliente: ")
    codigo = valida_codigo(verificar_existencia=True)

    cliente['nome'] = nome
    cliente['cidade'] = cidade
    cliente['codigo'] = codigo
    
    arquivo.write(f"{cliente['nome']}\t{cliente['cidade']}\t{cliente['codigo']}\n")
    arquivo.close()

def lista_clientes():
    """Lista os clientes do arquivo clientes.txt
    Args:
        None
    Returns:
        None
    """

    arquivo = open('clientes.txt', 'r')
    
    print("-" * 50)
    print(f"{'Nome':<20}{'Cidade':<20}{'Código':<20}")
    print("-" * 50)
    
    for linha in arquivo:
        resultado = linha.split("\t")
        print(f"{resultado[0]:<20}{resultado[1]:<20}{resultado[2]:<20}")
    print("-" * 50)
    a = input("Pressione ENTER para continuar...")
    arquivo.close()

def procura_codigo():
    """Procura um cliente pelo código no arquivo clientes.txt
    Args:
        None
    Returns:
        None
    """

    arquivo = open('clientes.txt', 'r')
    codigo = valida_codigo(verificar_existencia=False)

    busca = False
    
    cabeçalho("-", "Resultado da busca por código")
    
    for linha in arquivo:
        
        resultado = linha.split("\t")
        resultado[2] = resultado[2].replace("\n", "")
        
        if str(codigo) in resultado:
            busca = True
            print("-" * 50)
            print(f"{'Nome':<20}{'Cidade':<20}{'Código':<20}")
            print("-" * 50)
            print(f"{resultado[0]:<20}{resultado[1]:<20}{resultado[2]:<20}")
            break;
    print("-" * 50)
    
    if busca == False:
        print("Cliente não encontrado!")
        
    a = input("Pressione ENTER para continuar...")

    arquivo.close()
    
def procura_nome():
    """Procura um cliente pelo nome no arquivo clientes.txt
    Args:
        None
    Returns:
        None
    """
    arquivo = open('clientes.txt', 'r')
    nome = valida_string("Digite o nome do cliente: ")
    busca = False
    
    cabeçalho("-", "Resultado da busca por nome")
    
    for linha in arquivo:
        resultado = linha.split("\t")

        if nome in resultado[0]:
            busca = True
            print("-" * 50)
            print(f"{'Nome':<20}{'Cidade':<20}{'Código':<20}")
            print("-" * 50)
            print(f"{resultado[0]:<20}{resultado[1]:<20}{resultado[2]:<20}")
    print("-" * 50)
    
    if busca == False:
        print("Cliente não encontrado!")
        
    a = input("Pressione ENTER para continuar...")
    arquivo.close()

def verifica_codigo(codigo):
    """Procura um cliente pelo código no arquivo clientes.txt para verificar se já há algum com esse código.
    Args:
        codigo (int): codigo a ser verficado se já existe.
    Returns:
        bool: valor a ser retornado. True se já existir, False caso contrário.
    """

    arquivo = open('clientes.txt', 'r')

    busca = False
    
    for linha in arquivo:
        
        resultado = linha.split("\t")
        resultado[2] = resultado[2].replace("\n", "")

        if str(codigo) in resultado:
            busca = True
            break;
    
    return busca

def menu():
    """Menu principal do programa
    
    Args:
        None
    Returns:
        functon: menu (itself)
    
    """
    cabeçalho("=", "Menu principal")
    
    print("1|\t Adicionar cliente")
    print("2|\t Listar clientes")
    print("3|\t Procurar cliente por código")
    print("4|\t Procurar cliente por nome")
    print("5|\t Sair")
    print("-" * 45)
    
    opcao = input("Escolha uma opção: ")
    
    if opcao == "1":
        os.system(SISTEMA_OPERACIONAL)
        adiciona_cliente()
        sleep(1)
        os.system(SISTEMA_OPERACIONAL)
        return menu()
    elif opcao == "2":
        os.system(SISTEMA_OPERACIONAL)
        lista_clientes()
        os.system(SISTEMA_OPERACIONAL)
        return menu()
    elif opcao == "3":
        os.system(SISTEMA_OPERACIONAL)
        procura_codigo()
        os.system(SISTEMA_OPERACIONAL)
        return menu()
    elif opcao == "4":
        os.system(SISTEMA_OPERACIONAL)
        procura_nome()
        os.system(SISTEMA_OPERACIONAL)
        return menu()
    elif opcao == "5":
        print("Saindo...")
        sleep(1)
        os.system(SISTEMA_OPERACIONAL)
    else:
        print("Opção inválida!")
        sleep(1)
        os.system(SISTEMA_OPERACIONAL)
        return menu()